---
output:
  html_document:
    self_contained: no
---

```{r, include=F}
library(PalantiR)
library(knitr)
knitr::opts_chunk$set(cache=TRUE)
```

# Branch Heterogeneity

Branch heterogeneous simulations allow to specify different models (called `modes`) for each branch on a phylogeny. 
This is done by providing a newick tree with model indexes instead of branch length, and a list of models.

For more details on time-heterogeneous simulation, see [Temporal Heterogeneity Rescaling](Temporal_Heterogeneity_Rescaling.html).

Here is an example file, which represents a model switch on the primate lineage in the mammalian phylogeny:
```{r}
scan("../data/mammals_switch.newick", what = character(), quiet = T)
```

This phylogeny can be loaded with the `Phylogeny` constructor, with an additional `type = "mode"` argument.

```{r}
# Load phylogeny
mammals <- Phylogeny("../data/mammals.newick")

# Load branch modes
mammal_switches <- Phylogeny("../data/mammals_switch.newick", type = "mode")
```

We can view the phylogeny and the assigned modes:

```{r}
plot(mammals, mammal_switches)
```
# Example

In the example below, we simulate on a mammalian phylogeny with a population size expansion on the primate lineage.

## Setup

Load model parameters:

```{r}
# Load nucleotide equilibrium distribution
# Nucleotide order is T,C,A,G
nuc_pi <- c(.3, .2, .25, .25)

# Load amino acid equilibrium frequencies
# These will be converted to population-specific fitness vectors
aa_pi <- as.matrix(read.table("../data/psi-c50-1.txt"))

psi <- equilibrium_to_fitness(aa_pi[1, ], 1000)
```

Create models:

```{r}
# Create nucleotide substitution model
hky <- HasegawaKishinoYano(nuc_pi)

# Create codon models

# Small population size
ms_1k <- MutationSelection(
    population_size = 1000, 
    mutation_rate = 1e-8, 
    nucleotide_model = hky,
    fitness = psi)

# Large population size
ms_20k <- MutationSelection(
    population_size = 20000,
    mutation_rate = 1e-8, 
    nucleotide_model = hky,
    fitness = psi)

```

Sample sequence from model equilibrium distribution to start at root:
```{r sample_sequence}
s <- sample_sequence(ms_1k, 100)
```

Simulate:
```{r simulate}
sim <- simulate_over_interval_phylogeny(
    phylogeny = mammals,
    mode_phylogeny = mammal_switches,
    models = list(ms_1k, ms_20k),
    sequence = s,
    start_mode = 0)
```

Note that we have to specify which mode the root starts with using `start_mode` argument.

## Substitution history

Plot substitution history:

```{r plot_sim}
plot(sim, circle_opacity = 0.1)
```

Synonymous substiutions are shown in green, non-synonymous in red. Branch modes are shown as colored branches.
According to expectation, we observe a burst of non-synonymous substitutions at the start of the population size shift.
As time progresses, the number of non-synonymous substitutions goes down.

Check substitution history table:

```{r, results='hide'}
head(sim$substitutions)
```
```{r, echo=F, results='asis'}
kable(head(sim$substitutions))
```

## Alignment

Show alignment in a viewer:

```{r}
plot(sim$alignment)
```

To export `FASTA` sequence:

```{r, eval=F}
as.fasta(sim$alignment, file = "palantir_sim.fa")
```
