---
output: 
  html_document: 
    self_contained: no
---

```{r, include=F}
library(PalantiR)
library(knitr)
knitr::opts_chunk$set(fig.height = 7, fig.width = 9)
```

# Hasegawa-Kishino-Yano

Hasegawa Kishino Yano (HKY) [[1](https://www.ncbi.nlm.nih.gov/pubmed/3934395)] is a nucleotide substitution model which accounts for nucleotide equilibrium distribution rates of transitions ($\alpha$) and transversions($\beta$).

$$
Q_{ij} =
\begin{bmatrix}
    .            & \alpha \pi_C & \beta \pi_A  & \beta \pi_G  \\
    \alpha \pi_T & .            & \beta \pi_A  & \beta \pi_G  \\
    \beta \pi_T  & \beta \pi_C  & .            & \alpha \pi_G \\
    \beta \pi_T  & \beta \pi_C  & \alpha \pi_A & .            \\
\end{bmatrix}
$$
In `PalantiR` The HKY model is most commonly used to define nucleotide transition rates for Mutation Selection model, but it can also be used as a stand-alone simulation model.

# Example

Load model parameters:

```{r}
# Load phylogeny
mammals <- Phylogeny("../data/mammals.newick")

# Load nucleotide equilibrium distribution
# Nucleotide order is T,C,A,G
pi <- c(.3, .2, .25, .25)
```

Create model:

```{r}
hky <- HasegawaKishinoYano(
    equilibrium = pi,
    transition_rate = 2,
    transversion_rate = 1)
```

We first need to provide a sequence that to start off the simulation at the root.
We can sample a sequence from the equilibrium distribution of the model we are simulating:

```{r}
s <- sample_sequence(model = hky, length = 100)
```

Simulate:

```{r}
sim <- simulate_over_phylogeny(
    phylogeny = mammals, 
    model = hky, 
    sequence = s)
```

## Substitution history

Plot substitution history:

```{r}
plot(sim, sites = 0:10)
```

Transition substitutions are shown in green, transversions in red.

Check substitution history table:

```{r, results='hide'}
head(sim$substitutions)
```
```{r, echo=F, results='asis'}
kable(head(sim$substitutions))
```

We can change the colors on the plot by changing columns in the `substitutions` dataframe:
```{r}
transition_substitutions <- sim$substitutions$transition == TRUE
sim$substitutions$color[transition_substitutions] <- "orange"
sim$substitutions$color[!transition_substitutions] <- "blue"
plot(sim, sites = 0:10)
```

## Alignment

To export `FASTA` sequence:

```{r, eval=F}
as.fasta(sim$alignment, file = "palantir_sim.fa")
```

Show alignment in a scrollable viewer:

```{r}
plot(sim$alignment)
```

# References

1. Hasegawa, M., Kishino, H. and Yano, T. (1985). _Dating of the human-ape splitting by a molecular clock of mitochondrial DNA_. Journal of Molecular Evolution 22:160-174.



