---
output:
  html_document:
    self_contained: no
---

```{r, include=F}
library(PalantiR)
library(knitr)
knitr::opts_chunk$set(fig.height = 7, fig.width = 9)
```

# Shared Substitution Times

Shared substitution times is a switching simulation type similar to [Markov-Modulated Mutation-Selection](Markov_Modulated_Mutation_Selection.html). However, instead of simulating each site with independent switching in the parent model, the switch times are shared across sites. In essence, this is a conditioned simulation, given _times_ of parent model switches.

Unlike [Shared Model Substitution History](Shared_Substitution_History.html), this simulation does not share switch identity.

This type of simulation is useful to emulate fitness switches, such that fitness changes at every site at the same time. However, the identity of the fitness switch might differ.

# Example

## Setup

Load model parameters:

```{r}
# Load phylogeny
mammals <- Phylogeny("../data/mammals.newick")

# Load nucleotide equilibrium distribution
# Nucleotide order is T,C,A,G
nuc_pi <- c(.3, .2, .25, .25)

# Load amino acid equilibrium frequencies
# These will be converted to population-specific fitness vectors
aa_pi <- as.matrix(read.table("../data/psi-c50-1.txt"))

psi_1 <- equilibrium_to_fitness(aa_pi[1, ], 1000)
psi_2 <- equilibrium_to_fitness(aa_pi[2, ], 1000)
```

Create models:

```{r}
# Create nucleotide substitution model
hky <- HasegawaKishinoYano(nuc_pi)

# Parent switching model
gtr <- GeneralTimeReversible(
    equilibrium = rep(1/2, 2),
    exchangeability = matrix(rep(1/4, 4), nrow = 2))

# Create codon models

# First fitness vector
ms_1 <- MutationSelection(
    population_size = 1000, 
    mutation_rate = 1e-8, 
    nucleotide_model = hky,
    fitness = psi_1)

# Second fitness vector
ms_2 <- MutationSelection(
    population_size = 1000,
    mutation_rate = 1e-8, 
    nucleotide_model = hky,
    fitness = psi_2)

```

Sample sequence from model equilibrium distribution to start at root:
```{r sample_sequence}
s <- sample_sequence(ms_1, 25)
```

Simulate:
```{r simulate}
sim <- simulate_with_shared_time_heterogeneity(
    phylogeny = mammals,
    switching_model = gtr,
    switching_rate = 1,
    substitution_models = list(ms_1, ms_2),
    sequence = s,
    start_mode = 0)
```

This simulation will take considerably more time than any other type. The switch modes are not shared across sites, so the sampling matrices have to be rescaled for each site separately.

Note that we have to specify which mode the root starts with using `start_mode` argument.
We can additionally accelerate switching model substitution rate with `switching_rate`. 

## Substitution history

Plot substitution history:

```{r plot_sim}
plot(sim, circle_opacity = 0.1)
```

Synonymous substitutions are shown in green, non-synonymous in red. Since the switch types differ across sites, they are not shown on the phylogeny.
After every model switch, there is a burst of non-synonymous substitutions.

Check substitution history table:

```{r, results='hide'}
head(sim$substitutions)
```
```{r, echo=F, results='asis'}
kable(head(sim$substitutions))
```

Check model switching history table. Each site will have its own set of intervals. Each row corresponds to a tree node (branch). `state` is the index of the model, `from` and `to` are branch time points where the model applies:

```{r, results='hide'}
head(sim$intervals[[1]])
```
```{r, echo=F, results='asis'}
kable(head(sim$intervals[[1]]))
```

## Alignment

To export `FASTA` sequence:

```{r, eval=F}
as.fasta(sim$alignment, file = "palantir_sim.fa")
```

Show alignment in a scrollable viewer:

```{r}
plot(sim$alignment)
```

