---
output:
  html_document:
    self_contained: no
---

```{r, include=F}
library(PalantiR)
library(knitr)
knitr::opts_chunk$set(fig.height = 7, fig.width = 9)
```

# Shared Substitution History

Shared substitution history is a switching simulation type similar to [Markov-Modulated Mutation-Selection](Markov_Modulated_Mutation_Selection.html). However, instead of simulating each site with independent switching in the parent model, the switches are shared across sites. In essence, this is a conditioned simulation, given _identities_ and _times_ of parent model switches.

Unlike [Shared Model Time History](Shared_Substitution_Times.html), this simulation shares times and identities.

This type of simulation is useful to emulate population size switches, such that population size change is common for all sites in the alignment.

# Example

## Setup

Load model parameters:

```{r}
# Load phylogeny
mammals <- Phylogeny("../data/mammals.newick")

# Load nucleotide equilibrium distribution
# Nucleotide order is T,C,A,G
nuc_pi <- c(.3, .2, .25, .25)

# Load amino acid equilibrium frequencies
# These will be converted to population-specific fitness vectors
aa_pi <- as.matrix(read.table("../data/psi-c50-1.txt"))

psi <- equilibrium_to_fitness(aa_pi[1, ], 1000)
```

Create models:

```{r}
# Create nucleotide substitution model
hky <- HasegawaKishinoYano(nuc_pi)

# Parent switching model
gtr <- GeneralTimeReversible(
    equilibrium = rep(1/2, 2),
    exchangeability = matrix(rep(1, 4), nrow = 2))

# Create codon models

# Small population size
ms_1k <- MutationSelection(
    population_size = 1000, 
    mutation_rate = 1e-8, 
    nucleotide_model = hky,
    fitness = psi)

# Large population size
ms_20k <- MutationSelection(
    population_size = 20000,
    mutation_rate = 1e-8, 
    nucleotide_model = hky,
    fitness = psi)

```

Sample sequence from model equilibrium distribution to start at root:
```{r sample_sequence}
s <- sample_sequence(ms_1k, 100)
```

Simulate:
```{r simulate}
sim <- simulate_with_shared_substitution_heterogeneity(
    phylogeny = mammals,
    switching_model = gtr,
    switching_rate = 10,
    substitution_models = list(ms_1k, ms_20k),
    sequence = s,
    start_mode = 0)
```

Note that we have to specify which mode the root starts with using `start_mode` argument.
We can additionally accelerate switching model substitution rate with `switching_rate`. 

## Substitution history

Plot substitution history:

```{r plot_sim}
plot(sim, circle_opacity = 0.1)
```

Synonymous substitutions are shown in green, non-synonymous in red. Branch modes are shown as colored branches. Each mode has its own color. After every model switch, there is a burst of non-synonymous substitutions.

Check substitution history table:

```{r, results='hide'}
head(sim$substitutions)
```
```{r, echo=F, results='asis'}
kable(head(sim$substitutions))
```

Check model switching history table. Each row corresponds to a tree node (branch). `state` is the index of the model, `from` and `to` are branch time points where the model applies:

```{r, results='hide'}
head(sim$intervals)
```
```{r, echo=F, results='asis'}
kable(head(sim$intervals))
```

## Alignment

To export `FASTA` sequence:

```{r, eval=F}
as.fasta(sim$alignment, file = "palantir_sim.fa")
```

Show alignment in a scrollable viewer:

```{r}
plot(sim$alignment)
```

